### PR TITLE
Chores/unit test, feat/web3

### PR DESCRIPTION
**Pull request by sumitnitsurat**
- Added etherjs as a provider to interact with blockchain.
- Refactored code in react hooks 
- Unit test added for reducer and actions.
- static assets added.
- Document validation added.

### CODE REVIEW

File: src/components/home/actions/appActions.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import { appTypes } from "components/home/reducer/constants";
```
Comment (Nebulis) [inline src/components/home/actions/appActions.js]: typesafe-actions https://github.com/piotrwitek/typesafe-actions

File: src/components/home/actions/appActions.test.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import {
```
Comment (Nebulis) [inline src/components/home/actions/appActions.test.js]: to remove later => let's avoid doing sonar coverage

File: src/components/home/dropZone/index.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  const verificationStatus = getVerificationStatus(state);
+
+  useEffect(() => {
+    (async () => {
+      const wrapperDispatch = withMiddleware(state, dispatch)(updateNetworkId);
+      await wrapperDispatch(updateNetwork());
+    })();
+  }, [dispatch]); //passing state here causing the problem of rendering of document
+
+  const handleCertificateChange = certificate => {
```
Comment (Nebulis) [inline src/components/home/dropZone/index.js]: useCallback

File: src/components/home/dropZone/index.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import { updateNetworkId } from "components/home/utils/network";
+import DocumentDropZone from "./dropzone";
+import { Store } from "store";
+import { withMiddleware } from "middleware";
+import { verifyDocument } from "components/home/utils/validate";
+
+const DocumentDropZoneContainer = ({ history }) => {
+  const { state, dispatch } = React.useContext(Store);
+  const [fileError, setFileError] = useState(false);
+  const document = getDocument(state);
```
Comment (Nebulis) [inline src/components/home/dropZone/index.js]: next MR => useSelector ?

File: src/components/home/utils/network.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import { setNewWeb3, getWeb3, fetchNetwork } from "services/web3";
+import { appTypes } from "components/home/reducer/constants";
+import {
+  getNetworkPending,
+  getNetwork
+} from "components/home/actions/appActions";
+
+export const updateNetworkId = state => next => async action => {
+  try {
+    next(action);
```
Comment (Nebulis) [inline src/components/home/utils/network.js]: await

File: src/components/home/dropZone/views/defaultView.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
+import React from "react";
+import PropTypes from "prop-types";
+import "./viewerStyles.css";
+
+const View = ({ hover, accept, rootProps, inputProps }) => {
```
Comment (Nebulis) [inline src/components/home/dropZone/views/defaultView.js]: rename it

File: src/components/home/dropZone/views/defaultView.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+          <button type="button" className={`pointer ${"btn"}`}>
+            Select File
+          </button>
+          <input {...inputProps} />
+        </div>
+      </div>
+    </div>
+  );
+};
+export default View;
```
Comment (Nebulis) [inline src/components/home/dropZone/views/defaultView.js]: we might consider using named export

File: src/components/home/reducer/appReducer.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        networkUpdatePending: false
+      };
+    case appTypes.UPDATE_NETWORK_ID_FAILURE:
+      return {
+        ...state,
+        networkId: null,
+        networkIdVerbose: "",
+        networkUpdatePending: false
+      };
+    case appTypes.TX_POLLING_ADD:
```
Comment (Nebulis) [inline src/components/home/reducer/appReducer.js]: TRANSACTION or removed

File: src/components/home/reducer/appReducer.test.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
+import { appTypes, appInitialState } from "./constants";
```
Comment (Nebulis) [inline src/components/home/reducer/appReducer.test.js]: whatever is not there must be deleted

File: src/components/home/reducer/documentReducer.test.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import { docTypes, docInitialState } from "./constants";
```
Comment (Nebulis) [inline src/components/home/reducer/documentReducer.test.js]: whatever is not used must be removed

File: src/components/home/utils/network.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+export const getSelectedProvider = async (dispatch, getNew = false, store) => {
+  const networkPending = await getNetworkPending(store);
+  if (networkPending && !getNew) {
+    // block if there's a network update pending
+    await dispatch({ type: appTypes.UPDATE_NETWORK_ID_SUCCESS });
+  }
+  const network = await getNetwork(store);
+  const provider = (await getNew) ? setNewWeb3(network) : getWeb3(); // update web3 only if requested specifically
```
Comment (Nebulis) [inline src/components/home/utils/network.js]: need to be renamed, next MR ?

File: src/components/home/utils/network.test.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+describe("network", () => {
+  it("should update the network successfully", async () => {
+    const dispatch = jest.fn();
+    const action = {
+      type: appTypes.UPDATE_NETWORK_ID
+    };
+    const result = await updateNetworkId({ application: appInitialState })(
+      dispatch
+    )(action);
+    expect(result).toEqual(true);
```
Comment (Nebulis) [inline src/components/home/utils/network.test.js]: might consider testing side effect behaviour (global state)

File: src/components/home/utils/testUtils.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (Nebulis) [inline src/components/home/utils/testUtils.js]: please fix this one

File: src/components/home/utils/testUtils.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+  addIssuer(issuerString) {
+    const newIssuerObj = {
+      certificateStore: `71f10d54-d483-489b-b06f-fa2bed75ce16:string:${issuerString}`
+    };
+    this.cert.data.issuers.push(newIssuerObj);
+    return this;
+  }
+
+  finish() {
```
Comment (Nebulis) [inline src/components/home/utils/testUtils.js]: rename to build

File: src/components/home/utils/testUtils.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+export const mockStore = () => {
+  const stubbedFn = sinon.stub();
+  return {
+    isRevoked: stubbedFn,
+    isIssued: stubbedFn,
+    stub: stubbedFn
+  };
+};
+
+export class MakeCertUtil {
```
Comment (Nebulis) [inline src/components/home/utils/testUtils.js]: avoid shorter name => Cert is Certificate

File: src/components/home/utils/validate.js [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+import { docTypes } from "components/home/reducer/constants";
+import { getWeb3 } from "services/web3";
+const { info, error } = getLogger("util:validate");
+
+export async function loadCertificateContracts(payload, next) {
+  try {
+    const data = getData(payload);
+    info(`Loading document: ${data}`);
+
+    const unresolvedContractStoreAddresses = get(data, "issuers", []).map(
```
Comment (Nebulis) [inline src/components/home/utils/validate.js]: in typescript let's not use get => it's not type safe

File: src/components/home/utils/validate.js [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+import { getLogger } from "utils/logger";
+import DocumentStoreDefinition from "services/contracts/DocumentStore.json";
+// import fetchIssuers from "services/issuers";
+import { combinedHash } from "utils";
+import { ensResolveAddress } from "services/ens";
+import { docTypes } from "components/home/reducer/constants";
+import { getWeb3 } from "services/web3";
+const { info, error } = getLogger("util:validate");
+
+export async function loadCertificateContracts(payload, next) {
```
Comment (Nebulis) [inline src/components/home/utils/validate.js]: let's be functional :p

loadCertificateContracts = (payload) => (next) => {

File: src/reducers/index.js [hunk 18]
```diff
@@ -1,7 +1,13 @@
 import { combineReducers } from "redux";
 
-import document from "./document";
+import {
```
Comment (Nebulis) [inline src/reducers/index.js]: - [ ] file1
- [ ] file 2

File: src/services/ens/ens.js [hunk 19]
```diff
@@ -0,0 +1,114 @@
+import { stripHexPrefix, bufferToHex, keccak256 } from "ethereumjs-util";
```
Comment (Nebulis) [inline src/services/ens/ens.js]: ethereunNameService

File: src/services/issuers/index.js [hunk 20]
```diff
@@ -0,0 +1,9 @@
+const defaultUrl = "/static/registry.json";
+
+export default function fetchIssuers(url = defaultUrl) {
+  return window
```
Comment (Nebulis) [inline src/services/issuers/index.js]: window ?

File: src/components/home/actions/appActions.js (hunk 1 above)
Comment (sumitnitsurat) [reply src/components/home/actions/appActions.js]: created new issue.

File: src/components/home/dropZone/index.js (hunk 3 above)
Comment (sumitnitsurat) [reply src/components/home/dropZone/index.js]: created new issue.

File: src/components/home/dropZone/index.js (hunk 4 above)
Comment (sumitnitsurat) [reply src/components/home/dropZone/index.js]: created new issue https://github.com/TradeTrust/tradetrust-website/issues/24

File: src/components/home/dropZone/views/defaultView.js (hunk 6 above)
Comment (sumitnitsurat) [reply src/components/home/dropZone/views/defaultView.js]: Created new issue https://github.com/TradeTrust/tradetrust-website/issues/23

File: src/components/home/dropZone/views/defaultView.js (hunk 7 above)
Comment (sumitnitsurat) [reply src/components/home/dropZone/views/defaultView.js]: Created new issue https://github.com/TradeTrust/tradetrust-website/issues/23

File: src/components/home/reducer/appReducer.js (hunk 8 above)
Comment (sumitnitsurat) [reply src/components/home/reducer/appReducer.js]: Added in issue https://github.com/TradeTrust/tradetrust-website/issues/23

File: src/components/home/utils/network.js (hunk 11 above)
Comment (sumitnitsurat) [reply src/components/home/utils/network.js]: added in https://github.com/TradeTrust/tradetrust-website/issues/23

File: src/components/home/utils/network.test.js (hunk 12 above)
Comment (sumitnitsurat) [reply src/components/home/utils/network.test.js]: created new issue https://github.com/TradeTrust/tradetrust-website/issues/25

File: src/components/home/utils/testUtils.js (hunk 14 above)
Comment (sumitnitsurat) [reply src/components/home/utils/testUtils.js]: added in https://github.com/TradeTrust/tradetrust-website/issues/23

File: src/components/home/utils/testUtils.js (hunk 15 above)
Comment (sumitnitsurat) [reply src/components/home/utils/testUtils.js]: added in https://github.com/TradeTrust/tradetrust-website/issues/23

File: src/components/home/utils/validate.js (hunk 16 above)
Comment (sumitnitsurat) [reply src/components/home/utils/validate.js]: added in https://github.com/TradeTrust/tradetrust-website/issues/22

File: src/components/renderer/iframeRenderer/index.js [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+  const { state, dispatch } = React.useContext(Store);
+  const renderType = getRenderType(state);
+  const doc = getDocument(state);
+
+  useEffect(() => {
+    if (!doc) {
+      return props.history.push("/");
+    }
+
+    let iframe = document.querySelector("iframe");
```
[the remark this answers is not in the document]
Comment (sumitnitsurat) [reply src/components/renderer/iframeRenderer/index.js]: added in https://github.com/TradeTrust/tradetrust-website/issues/23

File: src/components/home/utils/validate.js [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+    await next({
+      type: docTypes.LOADING_STORE_SUCCESS
+    });
+    return contracts;
+  } catch (e) {
+    await next({
+      type: docTypes.LOADING_STORE_FAILURE,
+      payload: e
+    });
+    return null;
```
[the remark this answers is not in the document]
Comment (sumitnitsurat) [reply src/components/home/utils/validate.js]: added in https://github.com/TradeTrust/tradetrust-website/issues/23

File: src/components/home/utils/validate.js (hunk 17 above)
Comment (sumitnitsurat) [reply src/components/home/utils/validate.js]: https://github.com/TradeTrust/tradetrust-website/issues/23

**Pull request merged**